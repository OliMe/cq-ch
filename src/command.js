import getTransport from './event-transport/get-transport';
import { TYPE_COMMAND } from './constants';
import { checkArguments } from './helpers/argument-checker';
import { channelCreatorCfg, commandChannelCfg } from './arguments.cfg';

/**
 * Declares a channel for sent commands in the service interface.
 * @param {Array} types Types of commands for sending.
 * @param {string} context Application context e.g. Namespace of command.
 * @return {Function} Function for sending commands to channel.
 */
export default function command (types, context): Function {
  checkArguments([types, context], 'command', channelCreatorCfg);
  return command => {
    checkArguments([command], 'commandChannel', commandChannelCfg);
    if (!types.includes(command.type)) {
      throw new TypeError('Trying to send command with type not in interface.');
    }
    command = { ...command, context };
    getTransport(TYPE_COMMAND).trigger(command.type, command);
  };
}
