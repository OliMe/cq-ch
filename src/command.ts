import getTransport from './event-transport/get-transport';
import { TYPE_COMMAND } from './constants';
import { Context, Message, Types } from './types';
import { checkChannelCreator, checkCommandChannel } from './helpers/argument-checkers';

/**
 * Declares a channel for sent commands in the service interface.
 * @param {Array} types Types of commands for sending.
 * @param {string} context Application context e.g. Namespace of command.
 * @return {Function} Function for sending commands to channel.
 */
export default function command(types: Types, context: Context) {
  checkChannelCreator(command.name, types, context);
  return async function commandChannel(command: Message<undefined>) {
    checkCommandChannel(commandChannel.name, command);
    if (!types.includes(command.type)) {
      throw new TypeError('Trying to send command with type not in interface.');
    }
    const extendedCommand = { ...command, context };
    getTransport(TYPE_COMMAND).trigger(extendedCommand.type, extendedCommand);
  };
}
