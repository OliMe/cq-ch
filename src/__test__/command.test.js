import command from '../command';
describe('command', () => {
    it('creates commandChannel successfully', () => {
        expect(command(['test_type'], 'test')).toBeInstanceOf(Function);
    });
    it('creates commandChannel unsuccessfully', () => {
        const createChannelWithoutArguments = () => command()
        const createChannelWithFirstIncorrectArgument = () => command('not array')
        expect(createChannelWithoutArguments).toThrowErrorMatchingSnapshot();
        expect(createChannelWithFirstIncorrectArgument).toThrowErrorMatchingSnapshot();
    });
})
