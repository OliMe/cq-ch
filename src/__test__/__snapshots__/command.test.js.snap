// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`command creates commandChannel unsuccessfully 1`] = `"Failed to execute 'command': first argument must be an Array with at least one element."`;

exports[`command creates commandChannel unsuccessfully 2`] = `"Failed to execute 'command': first argument must be an Array with at least one element."`;
